framer/LayerStateMachine.coffee

```diff
@@ -1,51 +1,103 @@
-# {_} = require "./Underscore"
-#
-# {Events} = require "./Events"
 {BaseClass} = require "./BaseClass"
-# {Defaults} = require "./Defaults"
-
-{LayerStates} = require "./LayerStates"
 
 class exports.LayerStateMachine extends BaseClass
-	constructor: (layer) ->
+
+	constructor: (@_layer, @_states) ->
 		super
-		@_layer = layer
-		@properties = {}
-		@initial = LayerStates.filterStateProperties(layer.props)
+
 		@reset()
 
+	@define "layer",
+		get: -> @_layer
+
 	@define "current",
-		get: -> @states[@currentName]
+		get: -> @currentName
 
 	@define "previous",
-		get: -> @states[@previousName]
+		get: -> @previousName
+
+
+	@define "currentName",
+		get: -> @_currentName
 
 	@define "previousName",
 		get: -> _.last(@_previousNames)
 
 	@define "stateNames",
-		get: -> _.keys(@states)
+		get: -> Object.keys(@states)
```

Comment:
They are both equal, and I want to be sure `Object.keys` is the exact same as doing it on the layer.states.